packages/data-context/src/actions/LocalSettingsActions.ts

```diff
@@ -17,10 +18,8 @@ export class LocalSettingsActions {
     const toJson = JSON.parse(stringifiedJson) as AllowedState
 
     // update local data
-    for (const [key, value] of Object.entries(toJson)) {
-      this.ctx.coreData.localSettings.preferences[key as keyof AllowedState] = value as any
-    }
 
+    _.merge(this.ctx.coreData.localSettings.preferences, toJson)
```

Comment:
With nested objects in state now, fully replacing each key doesn't always make sense, so we merge instead to preserve any existing nested values not explicitly overridden by the client.